third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -637,17 +667,25 @@ impl<'a> FunctionGenerator<'a> {
         }
     }
 
-    /// Flush the abstract stack, ensuring that all values on the stack are stored in locals.
-    fn abstract_flush_stack(&mut self, ctx: &FunctionContext, top: usize) {
+    /// Flush the abstract stack, ensuring that all values on the stack are stored in locals, if
+    /// they are still alive.
+    fn abstract_flush_stack(&mut self, ctx: &BytecodeContext, top: usize) {
+        let fun_ctx = ctx.fun_ctx;
         while self.stack.len() > top {
             let temp = self.stack.pop().unwrap();
-            let local = self.temp_to_local(ctx, temp);
-            self.emit(FF::Bytecode::StLoc(local));
+            if ctx.is_alive_before(temp) || ctx.is_alive_after(temp) || self.pinned.contains(&temp)
```

Comment:
Why are you testing is_alive_before here?

Also, It seems that you only need to save a pinned local if it has been changed on the stack.  But abstract_push_result() already flushes a pinned temp.

We may need some targeted test cases that make this situation easier to see.

Reply:
When abstract_flush_stack is called, we are interested in the pre state. That is because the flushing will be essentially inserted _before_ the next instruction. This is different to when we use `is_alive_after` above: here we are interested in what is alive _after_ this given program point. FWIW this has been discovered by testing; before I would use the after-set.

Regards your 2nd point: There are temps on the stack which may in fact have never been stored into any local (as this is a stack machine -- in the ideal case we need to store as little as possible). Those need to be saved here. You are right we could optimize the case if the temp on the stack is loaded from a non-pinned local. I suggest we keep this optimization for later. Not really need the most optimal code here -- doing the liveness here so we can do meaningful ability and borrow analysis, which require the right placement of copy and move as pre-requisite.

Testing: please note this is quite extensively tested via the transactional tests [here](https://github.com/aptos-labs/aptos-core/tree/main/third_party/move/move-compiler-v2/transactional-tests/tests), which you do not see as part of this change. In fact I debugged quite a bit based on those tests.

Reply:
Regards the abstract stack, there is now also a section in the design doc: https://www.notion.so/aptoslabs/Aptos-Move-Compiler-V2-High-Level-Design-6ab5e5ffcd574c0693795e6fdefc1331?pvs=4#fc7f5dc9114a4a6097eeb75cdc2d34ff

Reply:
I think you should generate tight code based on the information you have available here, rather than being sloppy and assuming it will be fixed later.  We have the liveness results, let's use them effectively.

I'm afraid that abstract_flush_stack() is saving temps from the stack into locals in too many cases.  It might help to add comments for the pre/post conditions for a few of these shared operations, since the state of the function generator is a bit inconsistent as it is called in a few places at different points in generating code for a given input bytecode. 

The function is defined here: https://github.com/aptos-labs/aptos-core/blob/47803cd9d7ca46de87d9cfbbdd24678a59dbd144/third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs#L672.
and the overly-generous test is here:
https://github.com/aptos-labs/aptos-core/blob/47803cd9d7ca46de87d9cfbbdd24678a59dbd144/third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs#L676

(1) is_alive_before/after: See [function_generator.rs:152-156](https://github.com/aptos-labs/aptos-core/blob/47803cd9d7ca46de87d9cfbbdd24678a59dbd144/third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs#L152) (your branch) where we see:
```
                self.gen_bytecode(&bytecode_ctx, &bytecode[i], Some(next_bc));
                if !bc.is_branch() && matches!(next_bc, Bytecode::Label(..)) {
                    // At block boundaries without a preceding branch, need to flush stack
                    // TODO: to avoid this, we should use the CFG for code generation.
                    self.abstract_flush_stack(&bytecode_ctx, 0);
```
So in this particular case, **after** the bytecode is generated, we carefully store temps including those which are alive **before** the generated bytecode (even if not live **after** the bytecode).  In other cases (e.g., [in abstract_push_args()](https://github.com/aptos-labs/aptos-core/blob/47803cd9d7ca46de87d9cfbbdd24678a59dbd144/third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs#L656)) the flush is presumably happening **before** the current bytecode.  There's also a call [in abstract_push_result()](https://github.com/aptos-labs/aptos-core/blob/47803cd9d7ca46de87d9cfbbdd24678a59dbd144/third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs#L698) which also seems to happen **after** the bytecode is generated.

The current code test on line 676 (`if ctx.is_alive_before(temp) || ctx.is_alive_after(temp) || self.pinned.contains(&temp)`)  is conservative, and safe, in that it saves a superset of the required temps that works for both cases.  **But it would be better to differentiate the 2 cases (before and after the bytecode) so that spurious temp saves are not done.**

(2) is_pinned: The current code seems to ignore the lifetime of a pinned variable and consider it pinned for the rest of the method.  **Fine, though it would be good to mention it somewhere.**  Even with this conservative assumption, a pinned temp only needs to be saved if it has not already been copied to its home local.

There are only 2 locations where stack.push() is called.  (i) In abstract_push_args(), the temp is saved to a local before being pushed on the stack.  This case could presumably happen when pushing args for a Borrow operation, in which case the value would not yet have been saved, and should probably be saved to its home location immediately rather than waiting for abstract_flush_stack() to do it.

(ii) In abstract_push_result(), a temp is pushed, and if it is pinned, then it is saved right away, so there is no need to copy it later when abstract_flush_stack() is called and pops that temp.

So the is_pinned case is redundant in case (ii), and in case (i) it could be saved immediately.

(3) I don't see bytecode output for the transactional tests, so they aren't as useful to understand what is happening.

Reply:
Btw, since the current code is *safe*, I'd be ok with accepting this PR if we add an issue to go back and clarify the "stack temp needs to be stored in local" property in a latter PR.

Reply:
Ok.